core/capabilities/webapi/trigger.go

```diff
@@ -1,18 +1,293 @@
+	"A trigger to start workflow execution from a web api call",
+)
+
+type Input struct {
+}
+type Config struct {
+	AllowedSenders []string                 `toml:"allowedSenders"`
+	AllowedTopics  []string                 `toml:"allowedTopics"`
+	RateLimiter    common.RateLimiterConfig `toml:"rateLimiter"`
+	// RequiredParams is advisory to the consumer, it is not enforced.
+	RequiredParams []string `toml:"requiredParams"`
+}
+
+type webapiTrigger struct {
+	allowedSenders map[string]bool
+	allowedTopics  map[string]bool
+	ch             chan<- capabilities.TriggerResponse
+	config         Config
+	rateLimiter    *common.RateLimiter
+}
+
+type triggerConnectorHandler struct {
+	services.StateMachine
+
+	capabilities.CapabilityInfo
+	capabilities.Validator[Config, Input, capabilities.TriggerResponse]
+	connector           connector.GatewayConnector
+	lggr                logger.Logger
+	mu                  sync.Mutex
+	registeredWorkflows map[string]webapiTrigger
+}
+
+var _ capabilities.TriggerCapability = (*triggerConnectorHandler)(nil)
+var _ services.Service = &triggerConnectorHandler{}
+
+func NewTrigger(config string, registry core.CapabilitiesRegistry, connector connector.GatewayConnector, lggr logger.Logger) (*triggerConnectorHandler, error) {
+	if connector == nil {
+		return nil, errors.New("missing connector")
+	}
+	handler := &triggerConnectorHandler{
+		Validator:           capabilities.NewValidator[Config, Input, capabilities.TriggerResponse](capabilities.ValidatorArgs{Info: webapiTriggerInfo}),
+		connector:           connector,
+		registeredWorkflows: map[string]webapiTrigger{},
+		lggr:                lggr.Named("WorkflowConnectorHandler"),
+	}
+
+	return handler, nil
+}
+
+// processTrigger iterates over each topic, checking against senders and rateLimits, then starting event processing and responding
+func (h *triggerConnectorHandler) processTrigger(ctx context.Context, gatewayID string, body *api.MessageBody, sender ethCommon.Address, payload webapicapabilities.TriggerRequestPayload) {
+	// Pass on the payload with the expectation that it's in an acceptable format for the executor
+	wrappedPayload, err := values.WrapMap(payload)
+	if err != nil {
+		h.lggr.Errorw("Error wrapping payload")
+		err = h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "Error wrapping payload"})
+		if err != nil {
+			h.lggr.Errorw("error sending response", "err", err)
+		}
+		return
+	}
+	topics := payload.Topics
+
+	// empty topics is error for V1
+	if len(topics) == 0 {
+		h.lggr.Errorw("No Matching Workflow Topics")
+		err = h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "No Matching Workflow Topics"})
+		if err != nil {
+			h.lggr.Errorw("error sending response", "err", err)
+		}
+		return
+	}
+
+	matchedWorkflows := 0
+	var response webapicapabilities.TriggerResponsePayload
+	for _, trigger := range h.registeredWorkflows {
```

Comment:
Bump on https://github.com/smartcontractkit/chainlink/pull/14580#discussion_r1777867042

With the current code when multiple workflows with different allowed senders use the topics they will each fail out. This seems off. If that's the intention, maybe we should guard against this in register workflow?

Reply:
I'm not sure about that.  Imagine workflow A with topic T and AllowedSender S, workflow B with Topic T and AllowedSender S'.  There will be 2 triggers and they will have different allowedSenders, so I think the behavior codified is what we want.

Reply:
Imagine two workflows for different customers:

Workflow W1: for customer C1
```
- allowed senders: 0x1
- topics: "forex"
```

Workflow W2: for customer C2
```
- allowed senders: 0x2 
- topics: "forex"
```

User 0x2 sends in a request R1:
```
sender: 0x2
topic: "forex"
```

User 0x2 wants to trigger W2
Line 94 starts iterating every workflow, so starts on W1
Line 95 starts with the first topic on W1: "forex"
Line 96 passes because the topic of R1 is "forex"
Line 99 errors because the allowed sender of W1 is 0x1
User 0x2 has to supply a topic, but when they supply the topic "forex" this will fail.

Same ends up happening for W1 and user 0x1. The first iteration works as expected, but then the second matches on W2, sees that 0x1 is not an allowed sender, and errors.

Reply:
Yes, I see. I think we need a temporary error and keep proceeding.